app/services/commodities_tree_annotation_service.rb

```diff
@@ -0,0 +1,89 @@
+# Walk the commodities of a heading/subheading (assumed to be ordered correctly)
+# and annotate them with whether they are a leaf (have no children) and
+# with the id of their parent goods_nomenclature
+#
+# This is then consumed by the frontend tariff application to easily build a tree UI for perusal
+# of headings, subheadings and declarable commodities.
+class CommoditiesTreeAnnotationService
+  def initialize(heading)
+    @heading = heading
+  end
+
+  def call
+    traverse(@heading.commodities.first)
+
+    @heading
+  end
+
+  private
+
+  # TODO: Why can't we just iterate through a list of commodities in pairs and stamp them according to the conditions in map_goods_nomenclatures  method?
+  def traverse(current_goods_nomenclature)
+    # ignore case when first goods_nomenclature is blank it's a direct child of the heading
+    if @heading.commodities.index(current_goods_nomenclature).present?
+      next_goods_nomenclature = @heading.commodities[@heading.commodities.index(current_goods_nomenclature) + 1]
+
+      if next_goods_nomenclature.present? # we are not at the end of the goods_nomenclature array
+        map_goods_nomenclatures(current_goods_nomenclature, next_goods_nomenclature)
+
+        traverse(next_goods_nomenclature)
+      end
+    end
+  end
```

Comment:
I think we can just iterate, in my first attempt I wrote this:
```
def traverse(goods_nomenclatures, current_goods_nomenclature)
    current_good = current_goods_nomenclature

    while (index_gn = goods_nomenclatures.index(current_good))
      next_good = goods_nomenclatures[index_gn + 1]
      return unless next_good

      map_goods_nomenclatures(current_good, next_good)
      current_good = next_good
    end
  ```
  which is much simpler and can be improved further anyway.

Reply:
So the brief of this refactor was to extract/preserve original behaviour and add coverage.

Reply:
See this https://github.com/trade-tariff/trade-tariff-backend/pull/356/files#diff-fe5d0c8f58fc8ad73105d76099811869197f913b7728926b15a87016ebea9ebcR20

Reply:
At some point I will remove the recursion and iterate over staggered pairs of commodities